src/macros-by-example.adoc

```diff
@@ -0,0 +1,826 @@
+parameter accepts a value of any type.  But this is often undesirable, since the resulting output
+could be schematically malformed, or worse(?), macro-expansion could fail in hard-to-diagnose ways:
+
+[{mrk}]
+----
+(:detail_page_url [true]) ⇒ _**error**: make_string expects a string_
+----
+
+This E-expression cannot be expanded because `make_string` requires its arguments to be textual
+values, and `[true]` is not a string or symbol.  But this failure happens within the
+implementation of `detail_page_url`, not the point where the error occurred.  In this example
+those points are only one step removed, but it’s not hard to imagine macros where the stack is
+deep enough to make diagnosis quite difficult.
+
+To detect problems close to their source, macro signatures can declare type constraints on their parameters:
+
+[{mrk}]
+----
+(*public* detail_page_url
+  [(*string* asin)]
+  (website_url (make_string "dp/" asin)))
+----
+
+Here we constrain the `asin` parameter to produce a string so the blame is more clear:
+
+[{mrk}]
+----
+(:detail_page_url [true]) ⇒ _**error**: detail_page_url expects a string_
+----
+
+These types also serve a second, perhaps more impactful, purpose: they can allow the binary
+encoding to be more compact by omitting type information or using fixed-width values.
+
+This example reveals additional syntax for parameter declarations.  So far, a parameter was
+declared by a symbol denoting its name, now we have an S-expression pairing a type and a name.
+In this context the types include all the normal “concrete” Ion types, plus some useful abstract
+supertypes like `*number*`, `*text*`, and `*lob*`, as well as the unconstrained “top type” `*any*`.
+The latter is the default type, so the signature `[foo]` is equivalent to `[(*any* foo)]` meaning
+that the parameter `foo` accepts any value.
+
+
+=== Cardinality: Rest Parameters
+
+Some macros can accept an arbitrary number of arguments.  The `make_string` macro is one of those,
+concatenating zero or more arguments into a single string:
+
+----
+(:make_string)                 ⇒ ""
+(:make_string "a")             ⇒ "a"
+(:make_string "a" "b"    )     ⇒ "ab"
+(:make_string "a" "b" "c")     ⇒ "abc"
+(:make_string "a" "b" "c" "d") ⇒ "abcd"
+----
+
+To make this work, the definition of make_string is effectively:
+
+[{mrk}]
+----
+(*public* make_string [(*text \...* parts)] …)
+----
+
+This says that `parts` is a _rest parameter_ accepting zero or more arguments of type `*text*`.
+The `*\...*` modifier can only occur on the last parameter, declaring that “all the rest” of the
+arguments will be passed to that one name.
+
+NOTE: The Ion grammar treats identifiers like `text` and operators like `\...` as separate tokens
+even when juxtaposed like `text\...`. The template language syntax leverages that, and we’ll use the
+latter representation from now on.
+
+At this point our distinction between parameters and arguments starts showing its value, since
```

Comment:
*Minor*: Editorial suggestion--value is somewhat overloaded in this doc.

```suggestion
At this point our distinction between parameters and arguments becomes apparent, since
```